[Core][aDAG] Remove timeout for acquiring header for read/write

rkooo567 opened the pull request:
<!-- Thank you for your contribution! Please review https://github.com/ray-project/ray/blob/master/CONTRIBUTING.rst before opening a pull request. -->

<!-- Please add a reviewer to the assignee section when you create a PR. If you don't have the access to it, we will shortly find a reviewer and assign them to your PR. -->

## Why are these changes needed?

When users do read(timeout=0) or write(timeout=0), one may expect it returns immediately if the buffer is readable/writable. It is not true in some cases (I found it from https://github.com/ray-project/ray/pull/47272).

The root cause is that we apply timeout when we acquire headers. When a buffer is writable (its obj buffer is not acquired), it is possible write(timeout=0) fails if there are readers trying to acquire a header because we apply the same timeout when we acquire a header. This PR fixes the isuse by not applying the timeout in this case. 

It is okay because acquiring header takes a very short time, and it is immediately released. 

## Related issue number

<!-- For example: "Closes #1234" -->

## Checks

- [ ] I've signed off every commit(by using the -s flag, i.e., `git commit -s`) in this PR.
- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
    - [ ] I've added any new APIs to the API Reference. For example, if I added a 
           method in Tune, I've added it in `doc/source/tune/api/` under the 
           corresponding `.rst` file.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

rkooo567 commented on src/ray/object_manager/common.cc:
```diff
@@ -168,7 +168,7 @@ Status PlasmaObjectHeader::ReadAcquire(
   RAY_CHECK(sem.header_sem);
 
-  RAY_RETURN_NOT_OK(TryToAcquireSemaphore(sem.header_sem, timeout_point));
+  RAY_RETURN_NOT_OK(TryToAcquireSemaphore(sem.header_sem));
```
the below busy waiting TryToAcquireSemaphore needs timeout because we want to stop busy waiting when time is out. But this part doesn't need timeout.

ruisearch42 replied:
Could you add a comment why timeout is not applied here in the code, same for the other place

rkooo567 commented:
I could repro & confirmed fix from the buffering PR. But I had hard time finding standalone unit test for this case... suggestions are welcomed!

ruisearch42 commented on src/ray/object_manager/common.cc:
```diff
@@ -134,7 +134,7 @@ Status PlasmaObjectHeader::WriteAcquire(
 
   RAY_RETURN_NOT_OK(TryToAcquireSemaphore(sem.object_sem, timeout_point));
-  RAY_RETURN_NOT_OK(TryToAcquireSemaphore(sem.header_sem, timeout_point));
+  RAY_RETURN_NOT_OK(TryToAcquireSemaphore(sem.header_sem));
```
Should this be applied only when `timeout == 0`?

rkooo567 replied:
Hmm good question. I feel like in general header acquire time is very short (and it shouldn't take forever), so I feel like it is probably better having static timeout. For example, the same issue can happen when timeout is 0.0000001 (though it is practically not happening).

ruisearch42 commented on src/ray/object_manager/common.cc:
```diff
@@ -134,7 +134,7 @@ Status PlasmaObjectHeader::WriteAcquire(
 
   RAY_RETURN_NOT_OK(TryToAcquireSemaphore(sem.object_sem, timeout_point));
-  RAY_RETURN_NOT_OK(TryToAcquireSemaphore(sem.header_sem, timeout_point));
+  RAY_RETURN_NOT_OK(TryToAcquireSemaphore(sem.header_sem));
```
Can we have a unit test that fails without this change but passes with this change?

rkooo567 replied:
it is a bit difficult to write unit tests (I could verify it fixed issues before with bufferring PR when I used a different approach). Basically, there should be many downstream readers trying to acquire lock while a writer is trying to wrtie with timeout = 0. Let me try out

rkooo567 commented:
@ruisearch42 addressed comments. do you think it is necessary to add tests? I manually verified it fixes the issue in the past, and writing tests is a little tricky.
